pyphi/tpm.py

```diff
@@ -15,18 +15,262 @@
+                    if isinstance(result, tuple):
+                        # Multivalued "functions" result in a tuple, e.g., the
+                        # 2-tuple (quotients, remainders) from __divmod__().
+                        return (type(self)(r, validate=False) for r in result)
+
+                    return type(self)(result, validate=False)
+
+                return overriding_attribute
 
-class TPM:
-    """A transition probability matrix."""
+            return proxy
+
+        type.__init__(cls, type_name, bases, dct)
+
+        if cls.__wraps__:
+            ignore = cls.__ignore__
+            # Go through all the attribute strings in the wrapped array type.
+            for name in dir(cls.__wraps__):
+                # Filter special attributes. The rest will be handled by Wrapper.
+                if name.startswith("__") and name not in ignore and name not in dct:
+                    # Create proxy function for `name` and bind it to future
+                    # instances of cls.
+                    setattr(cls, name, property(make_proxy(name)))
+
+class Wrapper(metaclass=ProxyMetaclass):
+    """Proxy to the array inside PyPhi's custom TPM class.
+    """
+
+    __wraps__  = None
+    __ignore__ = {"class", "mro", "new", "init", "setattr", "getattr", "getattribute"}
+    __ignore__ = frozenset(f"__{i}__" for i in __ignore__)
 
     def __init__(self):
-        raise NotImplementedError
+        if self.__wraps__ is None:
+            raise TypeError("Base class Wrapper may not be instantiated.")
+
+        if not isinstance(self._tpm, self.__wraps__):
+            raise ValueError(f"Wrapped object must be of type {self.__wraps__}")
+
+
+class ExplicitTPM(Wrapper):
+    """An explicit network TPM in multidimensional form."""
+
+    __wraps__ = np.ndarray
+
+    # Casting semantics: values belonging to our custom TPM class should
+    # remain closed under the following methods:
+    __closures__ =  frozenset({
+        "argpartition", "astype", "byteswap", "choose", "clip", "compress",
+        "conj", "conjugate", "copy", "cumprod", "cumsum", "diagonal", "dot",
+        "fill", "flatten", "getfield", "item", "itemset", "max", "mean", "min",
+        "newbyteorder", "partition", "prod", "ptp", "put", "ravel", "repeat",
+        "reshape", "resize", "round", "setfield", "sort", "squeeze", "std",
+        "sum", "swapaxes", "take", "transpose", "var", "view"
+    })
+
+    # Proxy access to regular attributes of the wrapped array.
+    def __getattr__(self, name):
+        # Fix error with serialization. TODO: Implement dumps(), tobytes()?
+        if "_tpm" not in vars(self):
+            raise AttributeError
+
+        attribute = getattr(self._tpm, name)
+
+        if name not in self.__closures__:
+            return attribute
+
+        def overriding_method(*args, **kwargs):
+            # If second operand is a custom TPM, use its array instead.
+            if args and isinstance(args[0], type(self)):
```

Comment:
This logic looks like it duplicates the implementation in `ProxyMetaclass`; can we avoid that or is it necessary for some reason?

Reply:
Agreed. I first got both to work and only then I noticed the similarity.